src/rules/default-import-match-filename.js

```diff
@@ -0,0 +1,203 @@
+}
+
+/**
+ * Test if local name matches filename.
+ * @param {string} localName
+ * @param {string} filename
+ * @returns {boolean}
+ */
+function isCompatible(localName, filename) {
+  const normalizedLocalName = localName.replace(/_/g, '').toLowerCase()
+
+  return (
+    normalizedLocalName === normalizeFilename(filename) ||
+    normalizedLocalName === normalizeFilename(removeExtension(filename))
+  )
+}
+
+/**
+ * Match 'foo' but not 'foo/bar.js' and './foo'
+ * @param {string} path
+ * @returns {boolean}
+ */
+function isBarePackageImport(path) {
+  return path !== '.' && path !== '..' && !path.includes('/')
+}
+
+/**
+ * Match paths consisting of only '.' and '..', like '.', './', '..', '../..'.
+ * @param {string} path
+ * @returns {boolean}
+ */
+function isAncestorRelativePath(path) {
+  return (
+    path.length > 0 &&
+    !path.startsWith('/') &&
+    path
+      .split('/')
+      .every(segment => segment === '..' || segment === '.' || segment === '')
+  )
+}
+
+/**
+ * @param {string} packageJsonPath
+ * @returns {string | undefined}
+ */
+function getPackageJsonName(packageJsonPath) {
+  try {
+    const packageJsonContent = String(fs.readFileSync(packageJsonPath))
+    const packageJson = JSON.parse(packageJsonContent)
+    return packageJson.name || undefined
+  } catch (_) {
+    return undefined
+  }
+}
+
+function getNameFromPackageJsonOrDirname(path, context) {
+  const directoryName = Path.join(context.getFilename(), path, '..')
+  const packageJsonPath = Path.join(directoryName, 'package.json')
+  const packageJsonName = getPackageJsonName(packageJsonPath)
+  return packageJsonName || Path.basename(directoryName)
+}
+
+/**
+ * Get filename from a path.
+ * @param {string} path
+ * @param {object} context
+ * @returns {string | undefined}
+ */
+function getFilename(path, context) {
+  // like require('lodash')
+  if (isBarePackageImport(path)) {
+    return undefined
+  }
+
+  const basename = Path.basename(path)
+
+  const processedPath = /^index$|^index\./.test(basename)
+    ? Path.dirname(path)
+    : path
+
+  // like require('.'), require('..'), require('../..')
+  if (isAncestorRelativePath(processedPath)) {
+    return getNameFromPackageJsonOrDirname(processedPath, context)
+  }
+
+  return Path.basename(processedPath)
+}
+
+/**
+ * @param {string[]} ignorePaths
+ * @param {string} path
+ * @returns {boolean}
+ */
+function isIgnored(ignorePaths, path) {
+  return ignorePaths.some(pattern => path.includes(pattern))
+}
+
+module.exports = {
+  meta: {
+    type: 'suggestion',
+    docs: {
+      url: docsUrl('default-import-match-filename'),
+    },
+    schema: [
+      {
+        type: 'object',
+        additionalProperties: false,
+        properties: {
+          ignorePaths: {
+            type: 'array',
+            items: {
+              type: 'string',
+            },
+          },
+        },
+      },
+    ],
+  },
+
+  create(context) {
+    const ignorePaths = context.options[0]
```

Comment:
could this be made into a Set, so `has` operations on it are `O(1)`?

Reply:
There is no usage of `.has` for now, and there is usage of `.some`, so it is more convenient to leave it as an array.

Reply:
I’m not sure i understand why it’s more convenient, nor why existing usage of .has is relevant.